src/hll.c

```diff
@@ -92,6 +112,318 @@ enum {
     MST_UNINIT		= 0xffff,	// Internal uninitialized.
 };
 
+// Support disabling hash aggregation functionality for PG > 9.6
+#if PG_VERSION_NUM >= 90600
+
+#define HLL_EXTENSION_NAME "hll"
+#define HLL_ADD_AGG_NAME "hll_add_agg"
+#define HLL_UNION_AGG_NAME "hll_union_agg"
+
+typedef enum HllAggregationId {
+	HLL_ADD_AGG_ID = 0,
+	HLL_UNION_AGG_ID = 1
+} HllAggregationId;
+
+bool ForceGroupagg = false;
+static create_upper_paths_hook_type previous_upper_path_hook;
+static void RegisterHLLConfigVariables(void);
+#if (PG_VERSION_NUM >= 110000)
+static void hll_aggregation_restriction_hook(PlannerInfo *root, UpperRelationKind stage,
+								   RelOptInfo *input_rel, RelOptInfo *output_rel,
+								   void *extra);
+#else
+static void hll_aggregation_restriction_hook(PlannerInfo *root, UpperRelationKind stage,
+								   RelOptInfo *input_rel, RelOptInfo *output_rel);
+#endif
+static void MaximizeCostOfHashAggregateWithHLL(Path *path);
+static Oid LookupAggregatorOidByName(HllAggregationId hllAggregationId, int argumentCount);
+static Oid get_extension_schema(Oid ext_oid);
+static Oid FunctionOid(const char *schemaName, const char *functionName, int argumentCount, bool missingOk);
+static void RegisterHLLConfigVariables(void);
+
+/* HllAggregateCacheEntry contains information for each element in hllAggregaateCache */
+typedef struct HllAggregateCacheEntry
+{
+	Oid aggregateOid;
+	HllAggregationId hllAggregationId;
+	int argumentCount;
+} HllAggregateCacheEntry;
+
+static List *hllAggregateCache = NIL;
+
+void _PG_init(void);
+void _PG_fini(void);
+
+/* _PG_init is the shared library initialization function */
+void _PG_init(void)
+{
+	if (!process_shared_preload_libraries_in_progress)
+	{
+		ereport(ERROR, (errmsg("HLL can only be loaded via shared_preload_libraries"),
+						errhint("Add hll to shared_preload_libraries configuration "
+								"variable in postgresql.conf")));
+	}
+
+	/*
+	 * Register HLL configuration variables.
+	 */
+	RegisterHLLConfigVariables();
+
+	previous_upper_path_hook = create_upper_paths_hook;
+	create_upper_paths_hook = hll_aggregation_restriction_hook;
+}
+
+/*
+ * hll_aggregation_restriction_hook is assigned to create_upper_paths_hook to
+ * check whether there exist a path with hash aggregate. If that aggregate is
+ * introduced by hll, it's cost is maximized to force planner to not to select
+ * hash aggregate.
+ *
+ * Since the signature of the hook changes after PG 11, we define the signature
+ * and the previous hook call part of this function depending on the PG version.
+ */
+static void
+#if (PG_VERSION_NUM >= 110000)
+
+hll_aggregation_restriction_hook(PlannerInfo *root, UpperRelationKind stage,
```

Comment:
Curious, would it be possible to do this with [variable arguments](https://www.tutorialspoint.com/cprogramming/c_variable_arguments.htm) to save us from some `ifdefs`?